File: kattis/UFDS.java
import java.util.*;

class UFDS {
    private ArrayList<Integer> parents;
    private ArrayList<Integer> ranks;
    private ArrayList<Integer> size_of_set;
    private int num_of_set;

    public UFDS(int n) {
        this.parents = new ArrayList<>();
        this.ranks = new ArrayList<>();
        this.size_of_set = new ArrayList<>();
        this.num_of_set = n;

        for (int i = 0; i < n; i++) {
            this.parents.add(i);
            this.ranks.add(0);
            this.size_of_set.add(1);
        }
    }

    public int findSet(int x) { 
        if (parents.get(x) == x)  {
            return x;
        } else {
            int value = findSet(this.parents.get(x)); 
            parents.set(x, value);
            return value; 
        } 
    }

    public boolean isSameSet(int x, int y) { 
        return findSet(x) == findSet(y); 
    }

    public void unionSet(int m, int n) { 
        if (!isSameSet(m, n)) { 
            this.num_of_set--; 
            int x = findSet(m);
            int y = findSet(n);

            if (this.ranks.get(x) > this.ranks.get(y)) { 
                this.parents.set(y, x); 
                this.size_of_set.set(x, this.size_of_set.get(x) + this.size_of_set.get(y)); 
            } else { 
                this.parents.set(x, y); 
                this.size_of_set.set(y, this.size_of_set.get(y) + this.size_of_set.get(x));
                
                if (this.ranks.get(x) == this.ranks.get(y)) { 
                    this.ranks.set(y, this.ranks.get(y) + 1); 
                } 
            } 
        }
    }

    public int numOfDisjointSets() { 
        return this.num_of_set; 
    }
    
    public int sizeOfSet(int x) { 
        return this.size_of_set.get(findSet(x)); 
    }
}

File: python_library/UFDS.py
class UFDS:
	def __init__(self, n):
		self.parents = []
		self.ranks = []
		self.size_of_set = []
		self.num_of_set = n

		for i in range(n):
			self.parents.append(i)
			self.ranks.append(0)
			self.size_of_set.append(1)

	def find_set(self, x):
		if self.parents[x] == x:
			return x
		else:
			value = self.find_set(self.parents[x])
			self.parents[x] = value
			return value

	def is_same_set(self, x, y):
		return self.find_set(x) == self.find_set(y)

	def union_set(self, m, n):
		if not self.is_same_set(m,n):
			self.num_of_set -= 1
			x = self.find_set(m)
			y = self.find_set(n)

			if self.ranks[x] > self.ranks[y]:
				self.parents[y] = x
				self.size_of_set[x] += self.size_of_set[y]

			else:
				self.parents[x] = y
				self.size_of_set[y] += self.size_of_set[x]

				if self.ranks[x] == self.ranks[y]:
					self.ranks[y] += 1

	def num_of_disjoint_sets(self):
		return self.num_of_set

	def size_of_sets(self, x):
		return self.size_of_set[find_set(x)]
